src/validation.cpp

```diff
@@ -4909,7 +4952,14 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
 
     // Fake various pieces of CBlockIndex state:
     CBlockIndex* index = nullptr;
-    for (int i = 0; i <= snapshot_chainstate.m_chain.Height(); ++i) {
+
+    // Don't make any modifications to the genesis block.
+    // This is especially important because we don't want to erroneously
+    // apply BLOCK_ASSUMED_VALID to genesis, which would happen if we didn't skip
+    // it here (since it apparently isn't BLOCK_VALID_SCRIPTS).
```

Comment:
I don't understand this comment. If genesis isn't `BLOCK_VALID_SCRIPTS`, then `BLOCK_ASSUMED_VALID` wouldn't be applied, so the workaround isn't needed??

Reply:
No, *because* genesis isn't `BLOCK_VALID_SCRIPTS` we would apply `BLOCK_ASSUMED_VALID` unless we specifically skip it. The workaround is still desirable and necessary, I just bungled the unittest (see above).

Reply:
Yes, and the code below is:

```cpp
        if (!index->IsValid(BLOCK_VALID_SCRIPTS)) {
            // This flag will be removed once the block is fully validated by a
            // background chainstate.
            index->nStatus |= BLOCK_ASSUMED_VALID;
        }
```

So how could it be executed if `BLOCK_VALID_SCRIPTS` isn't set?

Reply:
It is only executed when `BLOCK_VALID_SCRIPTS` *is* set - are you by chance missing the `!` in the conditional?

Reply:
> Edit: changed an "is" to "isn't" - the double negative got me!

Did you push the change or am I missing it?

Reply:
Nvm, apparently I don't understand how negation works.

Reply:
This is an argument for including Python's `not` operator in C++21. ;)

Reply:
> This is an argument for including Python's `not` operator in C++21. ;)

It's actually already there, no need to wait https://en.cppreference.com/w/cpp/language/operator_alternative

Reply:
Pretty sure that `not` may fail on some compilers (on windows?).

In general I think if there is a way to avoid a negation, it might be preferable. In this case saying "(since it apparently is lacking BLOCK_VALID_SCRIPTS)" might work?